velox/exec/PrefixSort.cpp

```diff
@@ -0,0 +1,223 @@
+        prefix + prefixSortLayout.prefixOffset(index));
+  } else {
+    PrefixSortEncoder::encode(
+        *(reinterpret_cast<T*>(row + rowColumn.offset())),
+        prefix + prefixSortLayout.prefixOffset(index));
+  }
+  // Invert bits if desc
+  if (!prefixSortLayout.compareFlag(index).ascending) {
+    for (uint64_t s = 0; s < sizeof(T); s++) {
+      *(prefix + prefixSortLayout.prefixOffset(index) + s) =
+          ~*(prefix + prefixSortLayout.prefixOffset(index) + s);
+    }
+  }
+}
+} // namespace
+
+PrefixSortLayout::PrefixSortLayout(
+    const std::vector<TypePtr>& types,
+    const std::vector<CompareFlags>& keyCompareFlags,
+    const uint32_t maxPrefixKeyLength)
+    : normalizedKeySize_(0),
+      numNormalizedKeys_(0),
+      numSortKeys_(types.size()),
+      keyCompareFlags_(keyCompareFlags) {
+  VELOX_CHECK_GE(types.size(), 0);
+  for (TypePtr type : types) {
+    if (numNormalizedKeys_ > maxPrefixKeyLength) {
+      break;
+    }
+    if (type->kind() == TypeKind::BIGINT) {
+      numNormalizedKeys_++;
+      normalizedKeySize_ += sizeof(TypeTraits<TypeKind::BIGINT>::NativeType);
+    } else {
+      break;
+    }
+  }
+  entrySize_ = normalizedKeySize_ + sizeof(char*);
+  hasNonNormalizedKey_ = numNormalizedKeys_ < numSortKeys_;
+  // Compute prefix offsets for sort columns.
+  uint32_t offset = 0;
+  for (int i = 0; i < numNormalizedKeys_; i++) {
+    prefixOffsets_.push_back(offset);
+    offset += normalizeSize(types[i]->kind());
+  }
+}
+
+int PrefixSort::compareAllNormalizedKeys(char* left, char* right) {
+  // TODO : Support compare by words as a separate PR，benchmark performance
+  // could be enhanced by 2X.
+  return memcmp(left, right, sortLayout_.normalizedKeySize());
+}
+
+int PrefixSort::comparePartNormalizedKeys(char* left, char* right) {
+  int result = memcmp(left, right, sortLayout_.normalizedKeySize());
+  if (result != 0) {
+    return result;
+  }
+  // If prefixes are equal, compare the left sort keys with rowContainer.
+  char* leftAddress = getAddressFromPrefix(left);
+  char* rightAddress = getAddressFromPrefix(right);
+  for (int i = sortLayout_.numNormalizedKeys(); i < sortLayout_.numSortKeys();
+       i++) {
+    result = rowContainer_->compare(
+        leftAddress, rightAddress, i, sortLayout_.compareFlag(i));
+    if (result != 0) {
+      return result;
+    }
+  }
+  return 0;
+}
+
+PrefixSort::PrefixSort(
+    memory::MemoryPool* pool,
+    RowContainer* rowContainer,
+    const std::vector<CompareFlags>& keyCompareFlags,
+    const PrefixSortConfig& config)
+    : pool_(pool),
+      sortLayout_(
+          rowContainer->keyTypes(),
+          keyCompareFlags,
+          config.maxPrefixKeyLength),
+      rowContainer_(rowContainer) {}
+
+void PrefixSort::extractRowToPrefix(char* row, char* prefix) {
+  // Extract row to normalized prefix by column.
+  for (int32_t index = 0; index < sortLayout_.numNormalizedKeys(); index++) {
+    VELOX_DYNAMIC_TYPE_DISPATCH(
+        normalizeRowColumn,
```

Comment:
normalizeRowColumn seems very slow. I wonder how much time is spent copying data and normalizing the keys vs. sorting?

Reply:
1. complex is  : o(n) vs o(nlogn)
2. benchmark test show the cost < 5%,  I will detail it later.